File: day3/day3.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct claim
{
	int id;			/* identifier */
	int x;			/* x position */
	int y;			/* y position */
	int w;			/* width */
	int h;			/* height */

	struct claim *next;
};

int fabric[1000][1000];

static struct claim *load(FILE *input)
{
	struct claim *head = NULL;
	struct claim **tail = &head;
	struct claim c = {0};
	while (fscanf(input, " #%d @ %d,%d: %dx%d", &c.id, &c.x, &c.y, &c.w, &c.h) == 5)
	{
		struct claim *n = malloc(sizeof(*n));
		if (!n)
		{
			break;
		}
		*n = c;
		*tail = n;
		tail = &n->next;
	}
	return head;
}

static size_t part1(const struct claim *c)
{
	size_t count = 0;
	memset(fabric, 0, sizeof(fabric));
	for (; c; c = c->next)
	{
		for (int y = c->y; y < c->y+c->h; y++)
		{
			for (int x = c->x; x < c->x + c->w; x++)
			{
				fabric[y][x]++;
				if (fabric[y][x] == 2)
				{
					count++;
				}
			}
		}
	}
	return count;
}

static int part2(struct claim *c)
{
	/* part1 is a requisite */
	for(; c; c = c->next)
	{
		int found = 1;
		for (int y = c->y; found && y < c->y+c->h; y++)
		{
			for (int x = c->x; found && x < c->x+c->w; x++)
			{
				if (fabric[y][x] > 1)
				{
					found = 0;
				}
			}
		}
		if (found)
		{
			return c->id;
		}
	}
	return -1;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s filename\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "File %s not found\n", argv[1]);
		return 1;
	}

	struct claim *lst = load(input);
	fclose(input);
	if (!lst)
	{
		fprintf(stderr, "Cannot parse the data\n");
		return 1;
	}

	printf("Part1: %zu\n", part1(lst));
	printf("Part2: %d\n", part2(lst));
	while (lst)
	{
		struct claim *tmp = lst;
		lst = lst->next;
		free(tmp);
	}
	return 0;
}

File: day3/day3.py
#!/usr/bin/env python3

import re
import sys

def parse(txt):
    lst = []
    pattern = re.compile(r"^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$")
    for line in txt.splitlines():
        m = pattern.match(line)
        if m:
            lst.append(tuple(map(int, m.groups())))
    return lst

def project(claim):
    height = max(cx+cw for _,cx,_,cw,_ in claims)+1
    width = max(cy+ch for _,_,cy,_,ch in claims)+1
    fabric = [[0 for x in range(width)] for y in range(height)]
    for _, cx, cy, cw, ch in claims:
        for y in range(cy, cy+ch):
            for x in range(cx, cx+cw):
                fabric[y][x] += 1

    return fabric

def part1(fabric):
    return sum(sum(1 for val in row if val > 1) for row in fabric)

def part2(fabric, claims):
    for cid, cx, cy, cw, ch in claims:
        if all(fabric[y][x] == 1 for x in range(cx, cx+cw) for y in range(cy, cy+ch)):
            return cid
    return None

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]))
        sys.exit(1)

    claims = parse(txt)
    fabric = project(claims)
    print("Part1:", part1(fabric))
    print("Part2:", part2(fabric, claims))
